UGENE-6951. GUI tests for "Extract consensus from assemblies..." main menu item

evelinabis opened the pull request:
4 tests have been written to extract consensus from
1. .bam assembly
2. .bam.ugenedb assembly
3. multiple assemblies
4. invalid input

Commit:
UGENE-6951. GUI tests for "Extract consensus from assemblies..." main menu item.
4 tests have been written to extract consensus from
1. .bam assembly
2. .bam.ugenedb assembly
3. multiple assemblies
4. invalid input

Text scenarios have been added to "Common scenarios\Assembling\Extract
consensus"

yalgaer commented on src/plugins/GUITestBase/src/tests/common_scenarios/Assembling/extract_consensus/GTTestsAssemblyExtractConsensus.cpp:
```diff
@@ -0,0 +1,327 @@
+
+namespace GUITest_assembly_extract_consensus {
+
+struct ExtractConsensusWizardScenario : CustomScenario {
```
struct extends class: avoid this mix unless you have a reason for that.

evelinabis replied:
At the beginning, I thought that I would need to change the fields of the class/structure, in this regard, I did not find a reason to make unnecessary getters and setters, so I needed a structure with its public fields. However, later I realized that I could not reuse the same created object, because it is deleted, and therefore the need for the structure is gone. I'll fix that. Thank you for your comment.

yalgaer commented on src/plugins/GUITestBase/src/tests/common_scenarios/Assembling/extract_consensus/GTTestsAssemblyExtractConsensus.cpp:
```diff
@@ -0,0 +1,327 @@
+        SAMTOOLS
+    };
+
+    QStringList inpPaths;
```
nit: use the full name. Can't guess what "inp" means during the first read (write once, read 100 times)

yalgaer commented on src/plugins/GUITestBase/src/tests/common_scenarios/Assembling/extract_consensus/GTTestsAssemblyExtractConsensus.cpp:
```diff
@@ -0,0 +1,327 @@
+        : inpPaths(assembliesPath), algo(algo), keepGaps(keepGaps), outFile(outputFileName) {
+    }
+
+    void run(HI::GUITestOpStatus &os) {
```
nit: add override when you override.

yalgaer commented on src/plugins/GUITestBase/src/tests/common_scenarios/Assembling/extract_consensus/GTTestsAssemblyExtractConsensus.cpp:
```diff
@@ -0,0 +1,327 @@
+        CHECK_SET_ERR(dialog != nullptr, "activeModalWidget is nullptr");
+        {
+            auto *const urlleInp = GTWidget::findExactWidget<QLineEdit *>(os, "Assembly widget", dialog);
+            CHECK_SET_ERR(urlleInp != nullptr, "Assembly widget is nullptr");
```
nit: useless CHECK_SER_ERR: findExactWidget asserts for you unless you pass custom FindOptions that allow returning nullptr.
Same issue in many places below.

yalgaer commented on src/plugins/GUITestBase/src/tests/common_scenarios/Assembling/extract_consensus/GTTestsAssemblyExtractConsensus.cpp:
```diff
@@ -0,0 +1,327 @@
+            auto *const urlleInp = GTWidget::findExactWidget<QLineEdit *>(os, "Assembly widget", dialog);
+            CHECK_SET_ERR(urlleInp != nullptr, "Assembly widget is nullptr");
+            QString inpText;
+            foreach (const QString &inp, inpPaths) {
```
nit: foreach -> for. Same issue in other places below.

yalgaer commented on src/plugins/GUITestBase/src/tests/common_scenarios/Assembling/extract_consensus/GTTestsAssemblyExtractConsensus.cpp:
```diff
@@ -0,0 +1,327 @@
+                inpText += ";";
+            }
+            if (inpText.length() > 0) {
+                inpText.remove(inpText.length() - 1, 1);
```
nip: add a comment what is going on here. Maybe a more readable solution would be is to collect all paths into QStringList and to join then in the end: list.join(";"); (up to you)

yalgaer commented on src/plugins/GUITestBase/src/tests/common_scenarios/Assembling/extract_consensus/GTTestsAssemblyExtractConsensus.cpp:
```diff
@@ -0,0 +1,327 @@
+                                                << "Extract consensus from assemblies...");
+
+    //  4. Wait for workflow finished
+    GTUtilsTaskTreeView::waitTaskFinished(os, 100000);
```
nit: Default is 180*1000. May be you do not need to make an accent on 100*1000

yalgaer commented on src/plugins/GUITestBase/src/tests/common_scenarios/Assembling/extract_consensus/GTTestsAssemblyExtractConsensus.cpp:
```diff
@@ -0,0 +1,327 @@
+    //  Expected state: There should be no errors in the log
+    CHECK_SET_ERR(!lt.hasErrors(), "Errors in log: " + lt.getJoinedErrorString());
+    //  There should be no notifications in the dashboard
+    CHECK_SET_ERR(!GTUtilsDashboard::hasNotifications(os), "Unexpected notification");
```
nit: include the notification text in the assertion message to simplify analysis when the error will happen.

yalgaer commented on src/plugins/GUITestBase/src/tests/common_scenarios/Assembling/extract_consensus/GTTestsAssemblyExtractConsensus.cpp:
```diff
@@ -0,0 +1,327 @@
+        }
+
+        void run(HI::GUITestOpStatus &os) {
+            QWidget *const dialog = QApplication::activeModalWidget();
```
nit: 
Use GTWidget::getActiveModalWidget.
this method 1) Waits 2) Asserts that the model widget exists.

DmitriiSukhomlinov commented on src/plugins/GUITestBase/src/tests/common_scenarios/Assembling/extract_consensus/GTTestsAssemblyExtractConsensus.cpp:
```diff
@@ -0,0 +1,327 @@
+        QWidget *const dialog = QApplication::activeModalWidget();
+        CHECK_SET_ERR(dialog != nullptr, "activeModalWidget is nullptr");
+        {
+            auto *const urlleInp = GTWidget::findExactWidget<QLineEdit *>(os, "Assembly widget", dialog);
```
You have no need to write `auto *`, cuz `auto` could contain `*` "inside" itself. And, I know that it's OK, but we have used to text `const` before type

DmitriiSukhomlinov replied:
The same in other places

Commit:
UGENE-6951. GUI tests for "Extract consensus from assemblies..." main menu item
1. Completed multiple_inp_test
2. Removed duplication of the Algorithm enum

Commit:
UGENE-6951. GUI tests for "Extract consensus from assemblies..." main menu item
Changes by review:
1. Added a function to get a list of dashboard notifications (GTUtilsDashboard::getNotifications)
2. Replaced the abbreviated word "inp" with the full "input"
3. Replaced struct with a class for custom scenarios
4. Overridden methods are marked as overridden
5. Simplified run() methods of custom scenarios by GTWidget functions
6. Increased task waiting time
7. Notification messages are logged

yalgaer commented on src/plugins/GUITestBase/src/GTUtilsDashboard.cpp:
```diff
@@ -24,17 +24,78 @@
 
 #include "GTUtilsDashboard.h"
 
+namespace {
```
nit: you will save 4-5 vertical lines of code if you will use static instead of an anonymous namespace. However this is not a blocker, but only a suggestion. You can follow your style if you find it is better

yalgaer commented on src/plugins/GUITestBase/src/GTUtilsDashboard.cpp:
```diff
@@ -24,17 +24,78 @@
+QVariant getNotificationTypeFromString(const QString &str) {
+    using Notification = U2::GTUtilsDashboard::Notification;
+
+    QVariant returnable;
```
nit: normally you should use 'result' or 'notificationType'. It is the first time I see 'returnable' as the variable name. Usually, such names are used for types names.

yalgaer commented on src/plugins/GUITestBase/src/GTUtilsDashboard.cpp:
```diff
@@ -24,17 +24,78 @@
+    int counter = 0;
+    Notification::Type type;
+
+    for (int typeInd = 0; typeInd < typesNumber; ++typeInd) {
```
nit: every time you prefer ++i instead of i++ you want to emphasize that the returned value is important. Here it is not: I would suggest using more classic i++;

yalgaer replied:
It's just a style used across other code in the UGENE. You will find i++ to ++i as 100/1

yalgaer commented on src/plugins/GUITestBase/src/GTUtilsDashboard.cpp:
```diff
@@ -24,17 +24,78 @@
+    Notification::Type type;
+
+    for (int typeInd = 0; typeInd < typesNumber; ++typeInd) {
+        if (str.contains(Notification::typesNames[typeInd], Qt::CaseInsensitive)) {
```
Use startsWith not contains. Reason: a user text may contain the same text as the notification type name.

evelinabis replied:
str is string like `"<center><img src=":/U2Lang/images/error_20px.png"></center>"`, where a type is written in the png file name

yalgaer replied:
I see. In this case add 'class="warning|error|info"' marker to the notification HTML or use the existing class from the style-sheet (if this this available). 
In either case notification widget should have 100% info about notification type - no need to guess.

yalgaer commented on src/plugins/GUITestBase/src/GTUtilsDashboard.cpp:
```diff
@@ -24,17 +24,78 @@
+
+    QVariant returnable;
+    constexpr int typesNumber = sizeof(Notification::typesNames) / sizeof(QString);
+    int counter = 0;
```
if you use startsWith instead of contains you do not need the counter at all

yalgaer commented on src/plugins/GUITestBase/src/GTUtilsDashboard.cpp:
```diff
@@ -24,17 +24,78 @@
+
+#define GT_METHOD_NAME "getCellText"
+// Returns text from the (row, column) table cell. The (row, column) cell must exist
+QString getNotificationCellText(HI::GUITestOpStatus &os, const QGridLayout &tableLayout, const int row, const int col) {
```
class name is missed for the method. QString GTUtilsDashboard::getNotificationCellText

evelinabis replied:
It was a static function, not a class method. I will make this function a static method of the class

yalgaer commented on src/plugins/GUITestBase/src/GTUtilsDashboard.cpp:
```diff
@@ -24,17 +24,78 @@
+// Returns text from the (row, column) table cell. The (row, column) cell must exist
+QString getNotificationCellText(HI::GUITestOpStatus &os, const QGridLayout &tableLayout, const int row, const int col) {
+    QString text;
+    const QWidget *const cellWidget = tableLayout.itemAtPosition(row, col)->widget();
```
nit: too many local consts make the code less readable.

yalgaer replied:
Same here: you introduce very verbose style that is new and unusual to UGENE.
When it comes to local variables in a simple 10-line method 2 consts in variable type declaration makes the code less readable.

I see the following simple solutions:
- Want to avoid such comments: follow the common style found across other files in UGENE.
- Want to make this style to be a standard: write a email with a suggestion to the team alias before you use it.

In either case this comment is a 'nit' - not a blocker for the final approval

yalgaer commented on src/plugins/GUITestBase/src/GTUtilsDashboard.cpp:
```diff
@@ -24,17 +24,78 @@
+            }
+        }
+    }
+    return text;
```
Normally you should fail in the test method if the result is not found/available by default.

yalgaer commented on src/plugins/GUITestBase/src/GTUtilsDashboard.cpp:
```diff
@@ -24,17 +24,78 @@
+                                                              QStringLiteral("ERROR")};
+
+GTUtilsDashboard::Notification::operator QString() const {
+    const QString lBracket = QStringLiteral("[");
```
QStringLiterals are normally used to optimize the code. What kind of optimization you want to make here and why?
Use "["  or "]" to make the code more readable.

yalgaer commented on src/plugins/GUITestBase/src/GTUtilsDashboard.cpp:
```diff
@@ -146,6 +207,45 @@ bool GTUtilsDashboard::hasNotifications(HI::GUITestOpStatus &os) {
+QList<GTUtilsDashboard::Notification> GTUtilsDashboard::getNotifications(GUITestOpStatus &os) {
+    using Type = Notification::Type;
+
+    const QString notificationsWidgetName = QStringLiteral("NotificationsDashboardWidget");
```
Same here. Do not use string literals without special reason. Use QString notificationWidgetName = "";

yalgaer commented on src/plugins/GUITestBase/src/GTUtilsDashboard.cpp:
```diff
@@ -146,6 +207,45 @@ bool GTUtilsDashboard::hasNotifications(HI::GUITestOpStatus &os) {
+    using Type = Notification::Type;
+
+    const QString notificationsWidgetName = QStringLiteral("NotificationsDashboardWidget");
+    const auto tableLayout = qobject_cast<QGridLayout *>(
```
Use findExactWidget so you do not need to check for nullptr.

evelinabis replied:
It won't work here

yalgaer replied:
Ok, could you provide more details please why findExact won't work here?

evelinabis replied:
QLayout is not an inheritor of the QWidget class

I realized my mistake. I will make the code more readable

yalgaer commented on src/plugins/GUITestBase/src/GTUtilsDashboard.cpp:
```diff
@@ -146,6 +207,45 @@ bool GTUtilsDashboard::hasNotifications(HI::GUITestOpStatus &os) {
+        const QString typeStr = getNotificationCellText(os, *tableLayout, row, 0);
+        const QString element = getNotificationCellText(os, *tableLayout, row, 1);
+        const QString message = getNotificationCellText(os, *tableLayout, row, 2);
+        if (!typeStr.isNull() && !element.isNull() && !message.isNull()) {
```
use isEmpty instead of isNull for QStrings.

yalgaer commented on src/plugins/GUITestBase/src/GTUtilsDashboard.cpp:
```diff
@@ -146,6 +207,45 @@ bool GTUtilsDashboard::hasNotifications(HI::GUITestOpStatus &os) {
+QString GTUtilsDashboard::getJoinedNotificationsString(GUITestOpStatus &os) {
+    const auto notifications = getNotifications(os);
+    QStringList stringNotifications;
+    for (const auto &notification : notifications) {
```
use qAsConst in for loops explicitly

yalgaer commented on src/plugins/GUITestBase/src/GTUtilsDashboard.cpp:
```diff
@@ -146,6 +207,45 @@ bool GTUtilsDashboard::hasNotifications(HI::GUITestOpStatus &os) {
+    const auto notifications = getNotifications(os);
+    QStringList stringNotifications;
+    for (const auto &notification : notifications) {
+        stringNotifications << notification;
```
I would avoid operator overloading completely and called toString() here to make the code more readable and less dependable from the external context.

yalgaer commented on src/plugins/GUITestBase/src/GTUtilsDashboard.h:
```diff
@@ -59,7 +81,14 @@ class GTUtilsDashboard {
+
     static bool hasNotifications(HI::GUITestOpStatus &os);
+
+    static QList<Notification> getNotifications(HI::GUITestOpStatus &os);
```
For all new methods always add at least some docs.

yalgaer commented on src/plugins/GUITestBase/src/GTUtilsDashboard.h:
```diff
@@ -59,7 +81,14 @@ class GTUtilsDashboard {
+    static QList<Notification> getNotifications(HI::GUITestOpStatus &os);
+
+    // Returns a QString containing all dashboard notifications, splitted by '\n'
+    static QString getJoinedNotificationsString(HI::GUITestOpStatus &os);
```
This method is not needed at all: getNotificationList(os).join("\n") can be used instead

evelinabis replied:
What is `getNotificationList()`?

yalgaer replied:
I meant getNotifications() method. But you are right, there is no join(",") method there.

yalgaer commented on src/plugins/GUITestBase/src/tests/common_scenarios/Assembling/extract_consensus/GTTestsAssemblyExtractConsensus.cpp:
```diff
@@ -0,0 +1,318 @@
+#include "utils/GTUtilsDialog.h"
+
+namespace {
+enum class Algo {
```
Add docs for all new methods and types

yalgaer commented on src/plugins/GUITestBase/src/tests/common_scenarios/Assembling/extract_consensus/GTTestsAssemblyExtractConsensus.cpp:
```diff
@@ -0,0 +1,318 @@
+        QWidget *const dialog = GTWidget::getActiveModalWidget(os);
+
+        // Dialog filling
+        std::transform(inputPaths.begin(), inputPaths.end(), inputPaths.begin(), [](const QString &path) {
```
Use a simple indexed 'for'  loop - it will also take 3 lines but will be more readable/have fewer characters

yalgaer commented on src/plugins/GUITestBase/src/tests/common_scenarios/Assembling/extract_consensus/GTTestsAssemblyExtractConsensus.cpp:
```diff
@@ -0,0 +1,318 @@
+    GTToolbar::clickButtonByTooltipOnToolbar(os, MWTOOLBAR_ACTIVEMDI, "Show wizard");
+
+    //  7. Wait for workflow finished
+    GTUtilsTaskTreeView::waitTaskFinished(os, 180000);
```
2 timeouts by 3 minutes in the same test.
Nightly tests have timeout equal to 4 minutes in total.
Consider moving this test to the Metagenomics suite or reduce timeouts or test data to make the test faster.

yalgaer commented on src/plugins/GUITestBase/src/tests/common_scenarios/Assembling/extract_consensus/GTTestsAssemblyExtractConsensus.cpp:
```diff
@@ -0,0 +1,318 @@
+    GTToolbar::clickButtonByTooltipOnToolbar(os, MWTOOLBAR_ACTIVEMDI, "Show wizard");
+
+    //  7. Wait for workflow finished
+    GTUtilsTaskTreeView::waitTaskFinished(os, 180000);
```
Same here: 2 timeouts by 3 minutes

yalgaer commented on src/plugins/GUITestBase/src/tests/common_scenarios/Assembling/extract_consensus/GTTestsAssemblyExtractConsensus.cpp:
```diff
@@ -0,0 +1,318 @@
+    const QString dashboardErrMsg = "Unsupported document format: ";
+
+    const auto hasDashboardNotification = [](HI::GUITestOpStatus &os, const QString &errMsg) {
+        return !GTWidget::findLabelByText(os, errMsg, GTWidget::findWidget(
```
If you will cache results into the local variables you still will have only 3 lines. Split these lines into simpler independent statements.

evelinabis replied:
I don't understand what you are talking about

yalgaer replied:
Please cache GTWidget::findWidget() result into a separate variable to make the lines 171-173 more readable. I won't increase number of lines you have.

yalgaer commented on src/plugins/GUITestBase/src/GTUtilsDashboard.h:
```diff
@@ -39,6 +39,28 @@ class GTUtilsDashboard {
+                          WARN,
+                          ERR };
+
+        static const QString typesNames[3];
```
Rename to typeNames or typeNameList. typesNames (when both words are plural) is not correct.

Commit:
UGENE-6951. GUI tests for "Extract consensus from assemblies..." main
menu item
Changes by review:
1. The html-code of a picture in a dashboard notification has been
changed. Needed for testing
2. Minor changes in util function getNotifications
3. Reworked "Extract consensus" tests

Commit:
UGENE-6951. GUI tests for "Extract consensus from assemblies..." main menu item
Minor changes

yalgaer commented on src/plugins/GUITestBase/src/GTUtilsDashboard.cpp:
```diff
@@ -35,6 +37,33 @@ using namespace HI;
+
+    constexpr int typesNumber = sizeof(Notification::typeNames) / sizeof(QString);
+    for (int typeInd = 0; typeInd < typesNumber; typeInd++) {
+        if (QString::compare(str, Notification::typeNames[typeInd], Qt::CaseInsensitive) == 0) {
```
What is the reason to define typeName values in the upper-case and next compare them case insensitive with notification types which are lower-case?

I would suggest to use lowercase map and replace typeNames array with QStringList.
In this case your method can be as simple as 1-line of code.
 1. typeNames.indexOf(str).

yalgaer commented on src/plugins/GUITestBase/src/GTUtilsDashboard.cpp:
```diff
@@ -146,6 +175,80 @@ bool GTUtilsDashboard::hasNotifications(HI::GUITestOpStatus &os) {
+    const QString errMsg = "Dashboard notification type not found";
+    GT_CHECK_RESULT(!html.isEmpty(), errMsg, Notification::Type::ERR)
+
+    QXmlStreamReader reader(html);
```
I see big overdesign and non-needed complexity. Instead of using SAX parser (or DOM model) just locate the info type in the string like: html.indexOf(""<center><img class=\""")...
As the result this method may be 1-2 lines long.

yalgaer commented on src/plugins/GUITestBase/src/GTUtilsDashboard.cpp:
```diff
@@ -146,6 +175,80 @@ bool GTUtilsDashboard::hasNotifications(HI::GUITestOpStatus &os) {
+                    notifications)
+
+    const int notificationsNumber = tableLayout->rowCount();
+    notifications.reserve(notificationsNumber);
```
please remove optimization with reserve() unless you have a proof that this optimization is needed

yalgaer commented on src/plugins/GUITestBase/src/GTUtilsDashboard.h:
```diff
@@ -39,6 +39,32 @@ class GTUtilsDashboard {
+    // element to which the notification relates and the notification message
+    struct Notification {
+        // Type of notification
+        enum class Type { INFO,
```
Declaring Notification.Type as enum in the test package + writing a framework how to convert it to string & back looks like an over-design.

More than that the "type" used in the U2Designer code as a pure QString.

I would suggest to remove this enum and use plain old QString for the "type" in tests.
For the known set of types we can have a static named constants: static QString Notification::warning = "warning".

yalgaer commented on src/plugins/GUITestBase/src/GTUtilsDashboard.h:
```diff
@@ -39,6 +39,32 @@ class GTUtilsDashboard {
+        // if it is
+        static QVariant stringToType(const QString &str);
+
+        Notification() = delete;
```
Please document.

yalgaer commented on src/plugins/GUITestBase/src/GTUtilsDashboard.h:
```diff
@@ -39,6 +39,32 @@ class GTUtilsDashboard {
+        Notification() = delete;
+
+        Type type;
+        QString element;
```
Please document all new methods and fields.

yalgaer reviewed (changes requested):
Please schedule a "Nightly" run for this branch to check the new test results before approval.

Commit:
UGENE-6951. GUI tests for "Extract consensus from assemblies..." main
menu item
Minor changes by review

DmitriiSukhomlinov commented on src/plugins/GUITestBase/src/tests/common_scenarios/Assembling/extract_consensus/GTTestsAssemblyExtractConsensus.cpp:
```diff
@@ -0,0 +1,218 @@
+    //  1. Select "Tools->NGS data analysis->Extract consensus from assemblies..."
+    //  2. Set "samples/Assembly/chrM.sorted.bam" as an input
+    //  3. Click "Run"
+    auto *const bamScenario =
```
`auto` can define pointer by itself

Commit:
UGENE-6951. GUI tests for "Extract consensus from assemblies..." main
menu item
Fixed failure of Night testing, changes by review

yalgaer reviewed (commented):
The code looks OK, I have no other issues to fix. Please run Linux Nightly and when it shows now regressions I will approve.
